File: 프로그래머스/lv3/42861. 섬 연결하기/섬 연결하기.cpp
#include <bits/stdc++.h>

using namespace std;

int root[101];

int find(int x) {
    if(x == root[x])
        return x;
    
    return root[x] = find(root[x]);
}

void merge(int x, int y) {
    x = find(x);
    y = find(y);
    
    if(x == y)
        return;
    
    root[max(x, y)] = min(x, y);
}

int solution(int n, vector<vector<int>> costs) {
    int answer = 0;
    for(int i = 1 ; i <= n ; i++)
        root[i] = i;
    
    priority_queue < pair <int, pair<int, int>>, vector<pair <int, pair<int, int>>>, greater<> > pq;
    for(auto cost : costs) {
        pq.push({cost[2], {cost[0], cost[1]}});
    }
    while(!pq.empty()) {
        int curCost = pq.top().first;
        int curNodeA = pq.top().second.first;
        int curNodeB = pq.top().second.second;
        pq.pop();
        
        if(find(curNodeA) == find(curNodeB))         
            continue;
        
        merge(curNodeA, curNodeB);    
        answer += curCost;
        
        bool check = true;
        for(int i = 1 ; i <= n ; i++) {
            if(find(i) != find(1)) {
                check = false;
                break;
            }
        }
        if(check)
            break;
    }
    return answer;
}

File: 프로그래머스/3/42861. 섬 연결하기/섬 연결하기.js
const root = Array(101);

function find(x) {
    if(x == root[x])
        return x;
    
    return root[x] = find(root[x]);
}

function merge(x, y) {
    x = find(x);
    y = find(y);
    
    if(x == y)
        return;
    
    root[Math.max(x, y)] = Math.min(x, y);
}

function solution(n, costs) {
    var answer = 0;
    costs.sort((a, b) => a[2] - b[2]);
    
    for(let i = 0 ; i < n ; i++) {
        root[i] = i;
    }
    
    const s = new Set();
    for (const [nodeA, nodeB, cost] of costs) {
        if (find(nodeA) !== find(nodeB)) {
            
            merge(nodeA, nodeB);
            
            answer += cost;
            
            if(allConnected(n)) {
                break;
            }
        }
    }
    return answer;
}

function allConnected(n) {
    const tmp = root[0]; // 0번 노드 기준
    
    for (let i = 1; i < n; i++) {
        if (find(i) !== tmp) return false;
    }
    
    return true;
}